jax/experimental/export/_shape_poly_decision.py

```diff
@@ -276,27 +242,31 @@ def bounds(self, e: DimSize,
     if (const := _DimExpr.to_constant(e)) is not None:
       return (const, const)
     assert isinstance(e, _DimExpr)
-    # TODO: turn off the caching for now. Sometimes it leads to incompleteness.
-    # It is also too weak because if the stop_early is None, then the result
-    # should apply for any other stop_early. Similarly, if stop_early is
-    # _stop_early_for_geq0_or_leq0 and the result spans 0, then this should
-    # apply even if we don't have a stop_early.
-    #cache_key = (e, stop_early)
-    #if (res := self.scope._bounds_cache.get(cache_key)) is not None: return res
-    lb, ub = self._bounds_for_sorted_terms(e.scope, e._monomials_sorted, 0, stop_early)
-    res = (int(lb) if lb > -np.inf else lb,
-           int(ub) if ub < np.inf else ub)
-    #self.scope._bounds_cache[cache_key] = res
-    return res
+    # Caching bounds is tricky. Since the underlying _bounds_for_sorted_monomials
+    # is incomplete, and it may produce better results in the context of
+    # specific queries (due to the implicit constraints), if we cache the
+    # bounds computation we may stick to sub-optimal results. Also, we should
+    # not use the precision as part of the cache key, because a certain result
+    # may work for multiple precisions.
+    if (res := self.scope._bounds_cache.get(e)) is not None:
+      lb, ub, prev_prec = res
+      if _bounds_are_sufficient(prec, lb, ub): return (lb, ub)
+      if prev_prec.value >= prec.value: return (lb, ub)
+    lb, ub = self._bounds_for_sorted_terms(e.scope, e._monomials_sorted, 0, prec)
+    lb, ub = (int(lb) if lb > -np.inf else lb,
```

Comment:
Why do we need to cast `lb`/`ub` to `int` here?

Reply:
This is the entry point for the bounds computation and I want the results to be either infinity or integers. Most of the inner computations return integers already, but in some cases I may have missed some casts to integer, so I clean up here.